libexec/rtld-elf/aarch64/reloc.c

```diff
@@ -355,74 +324,552 @@ exclude_symbol_in_lib(const char *name, const char *sym, const Obj_Entry *obj, c
+			goto insert;
+		else
+#ifdef TRAMP_LINEAR_INSERTION
+			rtld_fatal("tramp_intern failed to insert key");
+#else
+			atomic_fetch_sub_explicit(&tramp_table.writers, 1,
+			    memory_order_relaxed);
+#endif
+	} while (key != target);
+	/*
+	 * Load-acquire the index until it becomes available.
+	 */
+	do
+	        idx = atomic_load_explicit(&tramp_table.map[slot].index,
+	            memory_order_acquire);
+	while (idx == -1);
+	entry = tramp_table.data[idx].entry;
+	goto end;
+insert:
+	idx = atomic_fetch_add_explicit(&tramp_table.size, 1,
+	    memory_order_relaxed);
+	/*
+	 * Invariant: tramp_table.size <= LOAD_FACTOR
+	 *
+	 * Construct the data array entry.
+	 */
+	entry = tramp_pgs_append(data);
+	tramp_table.data[idx] = *data;
+	tramp_table.data[idx].entry = entry;
+	/*
+	 * Store-release the index.
+	 */
+	atomic_store_explicit(&tramp_table.map[slot].index, idx,
+	    memory_order_release);
+	/*
+	 * If tramp_table.size == LOAD_FACTOR, resize the table.
+	 */
+	if (idx + 1 == tramp_table_load_factor(exp)) {
+#ifndef TRAMP_LINEAR_INSERTION
+		/*
+		 * Wait for other readers to complete.
+		 */
+		lock_upgrade(rtld_tramp_lock, &lockstate);
+#endif
+		/*
+		 * There can be no other writer racing with us for the resize.
+		 */
+		resizeTable(exp + 1);
+	}
+end:
+	lock_release(rtld_tramp_lock, &lockstate);
+	return (entry);
+}
+
+static void
+tramp_sig_validate(struct tramp_sig sig) {
+	if (sig.valid && sig.reg_args > 8)
+		rtld_fatal("Invalid tramp_sig.reg_args = %u", sig.reg_args);
+}
+
+static long
+tramp_sig_to_otype(struct tramp_sig sig) {
+	return sig.ret_args | (sig.mem_args << 2) | (sig.reg_args << 3);
+}
+
+void *
+_rtld_sandbox_code(void *target, struct tramp_sig sig)
+{
+	const Obj_Entry *obj;
+	void *target_unsealed;
+
+	tramp_sig_validate(sig);
+
+	if ((cheri_getperm(target) & CHERI_PERM_EXECUTIVE) != 0)
+		return (target);
+
+	obj = obj_from_addr(target);
+	if (obj == NULL)
+		rtld_fatal("%#p does not belong to any object", target);
+
+	target_unsealed = cheri_unseal(target, sealer_tramp);
+	if (cheri_gettag(target_unsealed)) {
+		if (sig.valid &&
+		    cheri_gettype(target) !=
+		    (long)cheri_getbase(sealer_tramp) + tramp_sig_to_otype(sig))
+			rtld_fatal("Signature mismatch");
+		target = cheri_sealentry(target_unsealed);
+	}
+
+	target = tramp_intern(&(struct tramp_data) {
+		.target = target,
+		.obj = obj,
+		.sig = sig
+	});
+
+	return (target);
+}
+
+void *
+_rtld_safebox_code(void *target, struct tramp_sig sig)
+{
+	const Obj_Entry *obj;
+
+	tramp_sig_validate(sig);
+
+	if ((cheri_getperm(target) & CHERI_PERM_EXECUTIVE) != 0)
+		return (target);
+
+	obj = obj_from_addr(target);
+	if (obj == NULL)
+		rtld_fatal("%#p does not belong to any object", target);
+
+	if (sig.valid) {
+		asm ("chkssu	%0, %0, %1\n"
+		    : "=C" (target)
+		    : "C" (obj->text_rodata_cap));
+		target = cheri_seal(target,
+		    sealer_tramp + tramp_sig_to_otype(sig));
+	}
+
+	return target;
 }
 
 #endif
```

Comment:
I'm guessing this ifdef matches `#if defined(__CHERI_PURE_CAPABILITY__) && defined(RTLD_SANDBOX)` far above. If so it would benefit from a comment.